src/inputs.rs

```diff
@@ -320,19 +346,26 @@ impl StatefulSource {
     fn next(&self, py: Python) -> PyResult<Poll<Option<TdPyAny>>> {
         match self.0.call_method0(py, "next") {
             Err(stop_ex) if stop_ex.is_instance_of::<PyStopIteration>(py) => Ok(Poll::Ready(None)),
-            Err(err) => Err(err),
+            Err(err) => Err(err).reraise("error getting next input in stateful source"),
```

Comment:
Is it possible to get the traceback into without needing to add a redundant message? Like https://github.com/bytewax/bytewax/pull/216/files#diff-316cbdd361ed963c0aa5d42dc4b0cb2dcb72b4a586da2bc4312164c5379709e9R218 will always be used whenever this error happens, so this extra annotation isn't super helpful noise.

Not a big deal if it's easier for everything to use the same code, though. One redundant message isn't a big deal.

Reply:
You are right that the reduntant message isn't helping. In fact in this specific case even the location isn't really useful, and I think it's ok to just remove this call to `.reraise`.

So right now we have this error:

```
Traceback (most recent call last):
  File "/home/docler/src/bytewax/bytewax/examples/errors/stateful_input_error.py", line 46, in <module>
    run_main(flow)
ValueError: (src/inputs.rs:218:34) error getting next input item from partition source
Caused by => ValueError: (src/inputs.rs:349:34) error getting next input in stateful source
Caused by => ValueError: A vague error
Traceback (most recent call last):
  File "/home/docler/src/bytewax/bytewax/examples/errors/stateful_input_error.py", line 20, in next
    raise ValueError("A vague error")
```

If we simply avoid adding a new message but still add location info, we'd get something like this (bar a different formatting of the error):

```
Traceback (most recent call last):
  File "/home/docler/src/bytewax/bytewax/examples/errors/stateful_input_error.py", line 46, in <module>
    run_main(flow)
ValueError: (src/inputs.rs:218:34) error getting next input item from partition source
Caused by => ValueError: (src/inputs.rs:349:34)
Caused by => ValueError: A vague error
Traceback (most recent call last):
  File "/home/docler/src/bytewax/bytewax/examples/errors/stateful_input_error.py", line 20, in next
    raise ValueError("A vague error")
```

But looking at the code, this function can only fail if the python code raises an exception.
Removing the `reraise` here and keeping it only at the caller site, we get this:
```

Traceback (most recent call last):
  File "/home/docler/src/bytewax/bytewax/examples/errors/stateful_input_error.py", line 46, in <module>
    run_main(flow)
ValueError: (src/inputs.rs:218:34) error getting next input item from partition source
Caused by => ValueError: A vague error
Traceback (most recent call last):
  File "/home/docler/src/bytewax/bytewax/examples/errors/stateful_input_error.py", line 20, in next
    raise ValueError("A vague error")
```

And if you go to the location, you'll notice that we just unwrapped a fallible function, and looking into it you can immediately see where it failed.
If this function was using `?` to propagate errors, or returning Errs in more than one way, it would make sense to reraise the exception here with different messages, but in this specific case it seems just noise.

So I'd keep the trait as it is unless we find other places where adding location without further info might still be useful (adding a method won't be much work given that they are automatically implemented by the trait),

Reply:
Sounds good to keep it as is.